### PR TITLE
SQL: introduce a query builder for the Rest tests

### PR DESCRIPTION
**Pull request by bpintea**
The new BaseRestSqlTestCase.RequestObjectBuilder class is a helper class
to build REST request objects for the tests. Consequently, "manual" string
concatenation to form JSON is done away with.
The class mimics SqlQueryRequestBuilder API.

The PR also extracts the attribute names used in SQL's JSON
request/replies as Protocol static final strings and updates the classes
that make use of the hard-coded strings to use these references.

### CODE REVIEW

File: x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/rest/BaseRestSqlTestCase.java [hunk 1]
```diff
@@ -19,10 +18,119 @@
...
+        public static RequestObjectBuilder cursor(String cursor) {
+            return new RequestObjectBuilder(field(CURSOR_NAME, cursor).substring(1), false);
+        }
+
+        public RequestObjectBuilder version(String version) {
+            request.append(field(VERSION_NAME, version));
+            return this;
+        }
+
+        public RequestObjectBuilder mode(String mode) {
```
Comment (costin) [inline x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/rest/BaseRestSqlTestCase.java:60]: I would add a method that accepts Mode directly so callers don't have to keep calling `mode(someMode.toString())`

**Review by costin — APPROVED** [review]
Thanks for looking at this. LGTM

File: x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/Protocol.java [hunk 2]
```diff
@@ -14,6 +14,32 @@
  * Sql protocol defaults and end-points shared between JDBC and REST protocol implementations
  */
 public final class Protocol {
+    /**
+     * The attribute names used in the protocol request/response objects.
+     */
+    // requests
+    public static final String QUERY_NAME = "query";
```
Comment (matriv) [inline x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/Protocol.java:21]: Nice!

File: x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/rest/BaseRestSqlTestCase.java [hunk 3]
```diff
@@ -19,10 +18,119 @@
...
+import static org.elasticsearch.xpack.sql.proto.Protocol.FILTER_NAME;
+import static org.elasticsearch.xpack.sql.proto.Protocol.MODE_NAME;
+import static org.elasticsearch.xpack.sql.proto.Protocol.PARAMS_NAME;
+import static org.elasticsearch.xpack.sql.proto.Protocol.QUERY_NAME;
+import static org.elasticsearch.xpack.sql.proto.Protocol.TIME_ZONE_NAME;
+
 public abstract class BaseRestSqlTestCase extends ESRestTestCase {
-    
+
+    public static class RequestObjectBuilder {
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/rest/BaseRestSqlTestCase.java:38]: Can this (and the methods) be private or package private?

**Review by matriv — APPROVED** [review]
LGTM. Very nice improvement!

File: x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/rest/BaseRestSqlTestCase.java (hunk 1 above)
Comment (bpintea) [reply x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/rest/BaseRestSqlTestCase.java]: Good idea, thanks.

File: x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/rest/BaseRestSqlTestCase.java (hunk 3 above)
Comment (bpintea) [reply x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/rest/BaseRestSqlTestCase.java:38]: They are called from outside the package (i.e. from `qa`) in a few instances.

**Review by astefan — APPROVED** [review]
LGTM

**Pull request merged**